### PR TITLE
Add extended key usage

### PR DESCRIPTION
**Pull request by garthy**
Fixes https://github.com/oxidecomputer/pki-playground/issues/3

This currently doesn't actually work and seems to fail if I add this to a cert 

See https://github.com/oxidecomputer/pki-playground/pull/4/commits/acf2550546f3fd6301bdebc5386bec156e7c45b8#diff-33af558753e6a6c431cf99fd8f11fb026226a5f6767c6e782749259c5d993f1e

```
Error: 
  × signing cert
  ╰─▶ message too long

```

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -90,6 +90,8 @@ pub struct Certificate {
 
     #[knuffel(child, unwrap(children))]
     pub extensions: Vec<X509Extensions>,
+
+
```
Comment (kc8apf) [inline src/config.rs:94]: Extraneous whitespace change.

File: src/config.rs [hunk 2]
```diff
@@ -147,6 +150,32 @@ pub struct KeyUsageExtension {
...
+    pub id_kp_email_protection: bool,
+
+    #[knuffel(child)]
+    pub id_kp_time_stamping: bool,
+
+    #[knuffel(child)]
+    pub id_kp_ocspsigning: bool,
+}
+
+
```
Comment (kc8apf) [inline src/config.rs:177]: Unnecessary newlines

File: src/lib.rs [hunk 3]
```diff
@@ -240,11 +242,16 @@ impl dyn Extension {
                 SubjectKeyIdentifierExtension::from_config(x, tbs_cert)?,
             )),
             config::X509Extensions::AuthorityKeyIdentifier(x) => Ok(Box::new(
-                AuthorityKeyIdentifierExtension::from_config(x, tbs_cert, issuer_cert)?,
```
Comment (kc8apf) [inline src/lib.rs:243]: Unnecessary whitespace change.

File: src/lib.rs [hunk 4]
```diff
@@ -240,11 +242,16 @@ impl dyn Extension {
...
+            AuthorityKeyIdentifierExtension::from_config(x, tbs_cert, issuer_cert)?,
             )),
+            config::X509Extensions::ExtendedKeyUsage(x) => Ok(Box::new(
+                ExtendedKeyUsageExtension::from_config(x)?),  
+            )
         }
     }
 }
+
```
Comment (kc8apf) [inline src/lib.rs:253]: Unnecessary whitespace change.

File: src/lib.rs [hunk 5]
```diff
@@ -334,6 +341,67 @@ impl KeyUsageExtension {
...
+    fn as_der(&self) -> &[u8] {
+        &self.der
+    }
+}
+
+
+impl ExtendedKeyUsageExtension {
+    pub(crate) fn from_config(config: &config::ExtendedKeyUsageExtension ) -> Result<Self> {
+
+        let mut der = Vec::new();
```
Comment (kc8apf) [inline src/lib.rs:369]:
```suggestion
        let mut key_purpose_ids = Vec::new();
```

**Review by kc8apf — COMMENTED** [review]
Looks correct.  A few style issues that rustfmt will probably correct for you.

**Comment by garthy** [comment]
Urgh sorry about formatting. So happy to get it compile have forgotten to configure local ide/tools to fmt

**Comment by kc8apf** [comment]
This does compile and appear to generate a correct certificate w/ the Extended Key Usage extension for me.  In fact, adding an extension _can't_ cause signing to fail as only a SHA-256 of the TbsCertificate is signed.  That failure is most likely an unrelated bug.  What platform are you using and does current main branch fail in the same way?  If so, please file another bug.

File: src/lib.rs (hunk 5 above)
Comment (garthy) [reply src/lib.rs:363]: I totally forgot to do this!

**Comment by kc8apf** [comment]
No worries. Can clean it up in a follow-up PR.
